compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala

```diff
@@ -629,6 +641,26 @@ trait BCodeSkelBuilder extends BCodeHelpers {
 
     } // end of method initJMethod
 
+    private def genTraitConstructorDefDef(dd: DefDef): Unit =
+      val statifiedDef = makeStatifiedDefDef(dd)
+      genDefDef(statifiedDef)
+
+    private def makeStatifiedDefDef(dd: DefDef): DefDef =
+      val origSym = dd.symbol.asTerm
+      val newSym = makeStatifiedDefSymbol(origSym, origSym.name)
+      tpd.DefDef(newSym, { paramRefss =>
+        val selfParamRef :: regularParamRefs = paramRefss.head
+        val enclosingClass = origSym.owner.asClass
+        new TreeTypeMap(
+          typeMap = _.substThis(enclosingClass, selfParamRef.symbol.termRef)
+            .subst(dd.vparamss.head.map(_.symbol), regularParamRefs.map(_.symbol.termRef)),
+          treeMap = {
+            case tree: This if tree.symbol == enclosingClass => selfParamRef
+            case tree => tree
+          }
+        ).transform(dd.rhs).changeOwner(origSym, newSym)
```

Comment:
TreeTypeMap has oldOwners/newOwners parameters, can't this be used instead of `changeOwner`? Example: https://github.com/lampepfl/dotty/blob/89871d82faa3d47e82e18dcfb6a35ca4ad70426d/compiler/src/dotty/tools/dotc/transform/FullParameterization.scala#L205-L215

Reply:
Yes, indeed, that's much better of course!